src/index/api.rs

```diff
@@ -0,0 +1,162 @@
+use std::io::Read;
+
+use crate::{error::Result, utils::MetadataPackage};
+
+use failure::err_msg;
+use reqwest::{header::ACCEPT, Client};
+use rustc_serialize::json::Json;
+use time::Timespec;
+
+pub(crate) struct RegistryCrateData {
+    pub(crate) release_time: Timespec,
+    pub(crate) yanked: bool,
+    pub(crate) downloads: i32,
+    pub(crate) owners: Vec<CrateOwner>,
+}
+
+pub(crate) struct CrateOwner {
+    pub(crate) avatar: String,
+    pub(crate) email: String,
+    pub(crate) login: String,
+    pub(crate) name: String,
+}
+
+impl RegistryCrateData {
+    pub(crate) fn get_from_network(pkg: &MetadataPackage) -> Result<Self> {
+        let (release_time, yanked, downloads) = get_release_time_yanked_downloads(pkg)?;
+        let owners = get_owners(pkg)?;
+
+        Ok(Self {
+            release_time,
+            yanked,
+            downloads,
+            owners,
+        })
+    }
+}
+
+/// Get release_time, yanked and downloads from the registry's API
+fn get_release_time_yanked_downloads(pkg: &MetadataPackage) -> Result<(time::Timespec, bool, i32)> {
+    let url = format!("https://crates.io/api/v1/crates/{}/versions", pkg.name);
```

Comment:
This URL won't work if, for example, we point the registry to staging.

Each index has a `config.json` file [has an optional `api` key](https://rust-lang.github.io/rfcs/2141-alternative-registries.html#registry-index-format-specification) that contains the URL to a crates.io-like API (see [crates.io's `config.json`](https://github.com/rust-lang/crates.io-index/blob/master/config.json) for an example). This struct should get the URL from the index config and use it as the prefix for network requests. If the `api` key is missing we should probably abort.

Reply:
Yep, I have another branch that handles all that, I just want to get it in piece by piece to avoid having a massive PR (and I still need to look into the `rustwide` changes necessary).

Reply:
Sounds good!